addons/addon-internal-auth-dep-check/packages/internal-auth-dep-check/lib/steps/internal-auth-dep-check-service.js

```diff
@@ -0,0 +1,255 @@
+    if (!(await this.verifyInternalUserWorkspacesAreTerminated(listOfInternalUsers))) {
+      throw this.boom.badRequest('Please terminate all internal user-owned workspaces before upgrading');
+    }
+
+    // Verify all SSH keys linked to internal users have been deleted
+    if (!(await this.verifyNoInternalUserSSHKey(listOfInternalUsers))) {
+      throw this.boom.badRequest('Please deactive all SSH Keys linked to internal users before upgrading');
+    }
+
+    // Verify internal users are not linked to the following resources: Projects, Org Studies, DS Accounts
+    if (!(await this.verifyNoInternalUserProjects(listOfInternalUsers))) {
+      throw this.boom.badRequest('Please deassociate all internal users from Projects before upgrading');
+    }
+    if (!(await this.verifyNoInternalUserOrgStudies(listOfInternalUsers))) {
+      throw this.boom.badRequest('Please deassociate all internal users from Organization Studies before upgrading');
+    }
+
+    // Verify all internal users (including root) are deactivated
+    if (!(await this.verifyNoActiveInternalUsers(listOfInternalStatuses))) {
+      throw this.boom.badRequest('Please deactive all Internal IdP users before upgrading');
+    }
+  }
+
+  async verifyInternalUserWorkspacesAreTerminated(listOfInternalUsers) {
+    const table = this.settings.get(settingKeys.environemntsTableName);
+
+    // Scan for any environments that are not in a failed or terminated state
+    const result = await this.dynamoDB
+      .scanner()
+      .table(table)
+      .names({ '#s': 'status' })
+      .values({ ':c': 'COMPLETED', ':p': 'PENDING', ':start': 'STARTING', ':stop': 'STOPPING', ':t': 'TERMINATING' })
+      .filter('#s = :c OR #s = :p OR #s = :start OR #s = :stop OR #s = :t')
+      .projection(['id', 'createdBy'])
+      .scan();
+
+    const listOfNonTerminatedWorkspaces = _.uniq(
+      result.filter(item => listOfInternalUsers.includes(item.createdBy)).map(item => item.id),
+    );
+
+    if (listOfNonTerminatedWorkspaces.length > 0) {
+      this.log.error(`Found at least one workspace that is not terminated: ${listOfNonTerminatedWorkspaces}`);
+      return false;
+    }
+
+    return true;
+  }
+
+  async verifyNoInternalUserSSHKey(listOfInternalUsers) {
+    const table = this.settings.get(settingKeys.keyPairsTableName);
+
+    // Scan for active ssh keys
+    const result = await this.dynamoDB
+      .scanner()
+      .table(table)
+      .names({ '#s': 'status' })
+      .values({ ':a': 'active' })
+      .filter('#s = :a')
+      .projection('uid')
+      .scan();
+
+    const listOfUsers = _.uniq(result.filter(item => listOfInternalUsers.includes(item.uid)).map(item => item.uid));
+
+    if (listOfUsers.length > 0) {
+      this.log.error(`Found at least one user with active SSH key: ${listOfUsers}`);
+      return false;
+    }
+    return true;
+  }
+
+  async _listInternalUsers() {
+    const table = this.settings.get(settingKeys.usersTableName);
```

Comment:
Just curious as to why we're not using our existing list-users API here.

Reply:
Because `listUsers` returns all users and here, I am leveraging the filters for DDB in the scan.